src/webextension/list/filter.js

```diff
@@ -7,29 +7,56 @@ export function parseFilterString(filter) {
   if (!filter) {
     return [];
   }
-  return filter.split(/\s+/).map((i) => i.toLocaleLowerCase());
+  return filter.split(/\s+/);
+}
+
+function tryMakeURL(string, defaultProtocol = null) {
```

Comment:
For what the user types, or the extension grabs from the activeTab (let's call this `target`), using exactly what is provided is appropriate.  For what we get from a saved entry in Lockbox (even if the entry had "http:") (let's call this `allowed`), I am far less sure.

Matching on "http:" carries a lot of risk.  It's ok to for "http://localhost' (or "http://127.0.0.1", or "http://[::1]"); this is considered a secure origin (at least for now).  For other hosts/ports, I worry we are potentially exposing users to exfiltration.  Also, matching on other origins carries risk, too (e.g., "javascript:").

This is probably a long way of saying the data massaging here needs to be more robust; my personal preference is sooner than later, but it'll need to happen.

\[I do see the valid protocol constraints, but they are far from the rest of this constraints logic]

Reply:
Where would we want this validation to happen? If we want to be strict about what's allowed in the datastore, should we just throw an exception when trying to save an entry if the user specifies a "bad" origin? At the moment, this code is assuming that the origin could be *literally anything*, since we have no constraints on what a user enters into the datastore. However, we could mandate that origins take a particular form, and possibly even normalize them when saving them (e.g. `www.google.com` is normalized to `https://www.google.com`). I'd actually prefer this to what I have now, since it's one less place where I have to make guesses about what the user meant.

For HTTP, how do we balance the fact that some sites just don't use HTTPS (and users would like to log into them) with the potential for some kind of MITM attack? Note that the code here only applies to the doorhanger, which is in chrome, not content. When we have a matcher used by (semi-) autofill, we may well choose to be even more restrictive. For example, I'd be ok with a plan where we say, "Autofill only works on HTTPS, but you can get pre-filtered results in the doorhanger even on HTTP".

I also think it would make sense to provide more restrictions here (e.g. constraining protocols), but we'd probably *also* have the constraints where they are now (or in a Redux action creator) since I don't think we should even fill in the search box on URLs with "bad" protocols. That said, a lot of this stuff will probably move to a common file somewhere once it's fleshed out a bit better.

Reply:
> Where would we want this validation to happen? If we want to be strict about what's allowed in the datastore, should we just throw an exception when trying to save an entry if the user specifies a "bad" origin? At the moment, this code is assuming that the origin could be *literally anything*, since we have no constraints on what a user enters into the datastore. However, we could mandate that origins take a particular form, and possibly even normalize them when saving them (e.g. `www.google.com` is normalized to `https://www.google.com`). I'd actually prefer this to what I have now, since it's one less place where I have to make guesses about what the user meant.

The datastore should have been more restrictive, and we can (should) make it more restrictive now (ish).  We still have to deal with existing data, though.

> For HTTP, how do we balance the fact that some sites just don't use HTTPS (and users would like to log into them) with the potential for some kind of MITM attack? Note that the code here only applies to the doorhanger, which is in chrome, not content. When we have a matcher used by (semi-) autofill, we may well choose to be even more restrictive. For example, I'd be ok with a plan where we say, "Autofill only works on HTTPS, but you can get pre-filtered results in the doorhanger even on HTTP".

Since _this_ code only applies to doorhanger's filter, I suppose _**for now**_ it's ok to match on `http:` also.  Howver, it is **MUST NOT** to extend that to our upcoming fill work, unless security-review finds it acceptable.